### PR TITLE
[RISCV] Add combines to form binop from tail insert idioms

### PR DESCRIPTION
**Pull request by preames**
This patch contains two related combines:
1) If we have an scalar vector insert into the result of a concat_vector,
   sink the insert into the operand of the concat.
2) If we have a insert of a scalar binop into a vector binop of the
   same opcode and the RHS of both are constant, perform the insert
   and then the binop.

The common theme to both is pushing inserts closer to the sources of the computation graph.  The goal is to enable forming vector bin ops from inserts of scalar binops at the end of another vector.

For RISCV specifically, the concat_vector transform will push inserts to smaller vectors.  This will have the effect of reducing lmul for the vslides, and usually doesn't require an additional vsetvli since
the source vectors are already working in the narrower VL.   I tried
that one as a target independent combine first, and it doesn't appear
profitable on all targets.

This is only one approach to the problem.  Another idea would be to aggressively form build_vectors and subvector inserts from the individual scalar inserts, and then have a transform which sunk a subvector_insert down through the concat.  The advantage of the alternate approach is that we expose parallelism in the insert sequence, even if the source vector isn't a concat_vector.  If reviewers are okay with it, I'd like to start with this approach, and then explore that direction in a follow up patch.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -14159,6 +14159,75 @@ static SDValue performBUILD_VECTORCombine(SDNode *N, SelectionDAG &DAG,
...
+      return SDValue();
+    // FIXME: Return failure if the RHS type doesn't match the LHS. Shifts may
+    // have different LHS and RHS types.
+    if (InVec.getOperand(0).getValueType() != InVec.getOperand(1).getValueType())
+      return SDValue();
+    SDValue LHS = DAG.getNode(ISD::INSERT_VECTOR_ELT, DL, VT,
+                              InVecLHS, InValLHS, EltNo);
+    SDValue RHS = DAG.getNode(ISD::INSERT_VECTOR_ELT, DL, VT,
+                              InVecRHS, InValRHS, EltNo);
+    return DAG.getNode(InVec->getOpcode(), DL, VT, LHS, RHS);
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:14200]: Nit
```suggestion
    return DAG.getNode(InVecOpcode, DL, VT, LHS, RHS);
```

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -14159,6 +14159,75 @@ static SDValue performBUILD_VECTORCombine(SDNode *N, SelectionDAG &DAG,
...
+    return SDValue();
+
+  if (!InVec.hasOneUse())
+    return SDValue();
+
+  // Given insert_vector_elt (binop a, VecC), (same_binop b, C2), Elt
+  // move the insert_vector_elts into the arms of the binop.  Note that
+  // the new RHS must be a constant.
+  const unsigned InVecOpcode = InVec->getOpcode();
+  if (InVecOpcode == InVal->getOpcode() && TLI.isBinOp(InVecOpcode) &&
```
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:14181]: I think this combine would also work on scalable vectors too, is it worthwhile moving it up before the scalable vector check?

Comment (preames) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:14181]: I believe you're right.  I'd prefer to leave this as is for this change as the interaction of the combines is rather fiddly.  Happy to revisit in a follow up change though.

**Review by lukel97 — APPROVED** [review]
LGTM other than clang-format warnings

**Pull request merged**